File: lab4/AutomationScenario.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SmartHomeSystem
{
    public class AutomationScenario
    {
        public string ScenarioId { get; private set; }
        public string Name { get; set; }
        public string TriggerCondition { get; set; }
        private List<ScenarioAction> actions;
        public bool IsActive { get; set; }
        public DateTime CreatedDate { get; private set; }

        public AutomationScenario(string scenarioId, string name, string triggerCondition)
        {
            ScenarioId = scenarioId;
            Name = name;
            TriggerCondition = triggerCondition;
            actions = new List<ScenarioAction>();
            IsActive = true;
            CreatedDate = DateTime.Now;
        }

        public void AddAction(ScenarioAction action)
        {
            if (!actions.Contains(action))
            {
                actions.Add(action);
            }
        }

        public void RemoveAction(ScenarioAction action)
        {
            actions.Remove(action);
        }

        public List<ScenarioAction> GetActions()
        {
            return new List<ScenarioAction>(actions);
        }

        public int GetActionCount()
        {
            return actions.Count;
        }

        public void Activate()
        {
            IsActive = true;
            Console.WriteLine($"[SCENARIO] Сценарий '{Name}' активирован");
        }

        public void Deactivate()
        {
            IsActive = false;
            Console.WriteLine($"[SCENARIO] Сценарий '{Name}' деактивирован");
        }

        public void Execute()
        {
            if (!IsActive)
            {
                Console.WriteLine($"[SCENARIO] Сценарий '{Name}' неактивен");
                return;
            }

            Console.WriteLine($"[SCENARIO] Выполнение сценария '{Name}'...");
            foreach (var action in actions)
            {
                action.Execute();
            }
            Console.WriteLine($"[SCENARIO] Сценарий '{Name}' выполнен");
        }

        public void DisplayInfo()
        {
            Console.WriteLine($"┌────────────────────────────────────────");
            Console.WriteLine($"│ Сценарий: {Name}");
            Console.WriteLine($"│ ID: {ScenarioId}");
            Console.WriteLine($"│ Условие: {TriggerCondition}");
            Console.WriteLine($"│ Статус: {(IsActive ? "АКТИВЕН" : "НЕАКТИВЕН")}");
            Console.WriteLine($"│ Действий: {actions.Count}");
            Console.WriteLine($"└────────────────────────────────────────");
        }

        public void DisplayActions()
        {
            Console.WriteLine($"\nДействия сценария '{Name}':");
            if (actions.Count == 0)
            {
                Console.WriteLine("  Действий нет.");
                return;
            }

            foreach (var action in actions)
            {
                action.DisplayInfo();
            }
        }

        public string Serialize()
        {
            string actionIds = string.Join(",", actions.Select(a => a.ActionId));
            return $"{ScenarioId}|{Name}|{TriggerCondition}|{IsActive}|{actionIds}";
        }

        public static AutomationScenario Deserialize(string data)
        {
            var parts = data.Split('|');
            if (parts.Length != 5) return null;

            return new AutomationScenario(
                parts[0],
                parts[1],
                parts[2]
            )
            {
                IsActive = bool.Parse(parts[3])
            };
        }

        public override string ToString()
        {
            return $"{Name} ({actions.Count} действий)";
        }
    }
}

File: smarthomec++/AutomationScenario.cpp
#include "automationScenario.hpp"
#include "scenarioAction.hpp"
#include <iostream>
#include <iomanip>
#include <sstream>

AutomationScenario::AutomationScenario(const std::string& id, const std::string& scenarioName,
    const std::string& time)
    : scenarioId(id), name(scenarioName), triggerTime(time),
    isActive(false), createdDate(std::time(nullptr)) {}

AutomationScenario::~AutomationScenario() {
    actions.clear();
}

void AutomationScenario::activate() {
    isActive = true;
    std::cout << "Сценарий '" << name << "' активирован" << std::endl;
}

void AutomationScenario::deactivate() {
    isActive = false;
    std::cout << "Сценарий '" << name << "' деактивирован" << std::endl;
}

void AutomationScenario::execute() {
    if (!isActive) {
        std::cout << "Сценарий '" << name << "' не активен" << std::endl;
        return;
    }
    std::cout << "Выполнение сценария: " << name << std::endl;
    for (auto& action : actions) {
        action->execute();
    }
}

void AutomationScenario::addAction(std::unique_ptr<ScenarioAction> action) {
    actions.push_back(std::move(action));
}

void AutomationScenario::removeAction(ScenarioAction* action) {
    for (auto it = actions.begin(); it != actions.end(); ++it) {
        if (it->get() == action) {
            actions.erase(it);
            break;
        }
    }
}

void AutomationScenario::displayInfo() const {
    std::cout << "Сценарий: " << name << " (" << (isActive ? "Активен" : "Неактивен") << ")\n";
    std::cout << "Время запуска: " << triggerTime << "\n";
    std::cout << "Действий: " << actions.size() << "\n";
}

void AutomationScenario::displayActions() const {
    std::cout << "Действия сценария '" << name << "':\n";
    for (size_t i = 0; i < actions.size(); i++) {
        std::cout << "  " << i + 1 << ". ";
        actions[i]->displayInfo();
    }
}

std::string AutomationScenario::getScenarioId() const {
    return scenarioId;
}

std::string AutomationScenario::getName() const {
    return name;
}

std::string AutomationScenario::getTriggerTime() const {
    return triggerTime;
}

bool AutomationScenario::getIsActive() const {
    return isActive;
}

int AutomationScenario::getActionCount() const {
    return actions.size();
}

std::string AutomationScenario::serialize() const {
    std::stringstream ss;
    ss << scenarioId << "|" << name << "|" << triggerTime << "|" << isActive << "|" << createdDate;
    return ss.str();
}

std::unique_ptr<AutomationScenario> AutomationScenario::deserialize(const std::string& data) {
    std::stringstream ss(data);
    std::string id, name, time, activeStr, dateStr;
    std::getline(ss, id, '|');
    std::getline(ss, name, '|');
    std::getline(ss, time, '|');
    std::getline(ss, activeStr, '|');
    std::getline(ss, dateStr, '|');
    auto scenario = std::make_unique<AutomationScenario>(id, name, time);
    if (activeStr == "1") scenario->activate();
    return scenario;
}